tools/fdleak.py

```diff
@@ -0,0 +1,329 @@
+    return 0;
+
+  pidfd += (u32)ret_fd;
+  info.timestamp_ns = bpf_ktime_get_ns();
+  info.stack_id =
+      user_stack.get_stackid(ctx, BPF_F_USER_STACK | BPF_F_REUSE_STACKID);
+  info.tid = tid;
+  info.fd = ret_fd;
+  bpf_get_current_comm(info.comm, TASK_COMM_LEN);
+
+  fd_allocs_hash.update(&pidfd, &info);
+  update_statistics_add(info.stack_id);
+
+  ##PRINT_DEBUG##
+  bpf_trace_printk("[%s-%u] alloc fd[%d] entered \\n",
+                    info.comm, info.tid, ret_fd);
+
+  return 0;
+}
+
+static inline void update_statistics_del(int stack_id, u64 survivor_time_ns) {
+  struct combined_alloc_info_t *existing_cinfo;
+  struct combined_alloc_info_t cinfo = {0};
+
+  existing_cinfo = combined_allocs_hash.lookup(&stack_id);
+  if (existing_cinfo == NULL) {
+    bpf_trace_printk("[ERROR] missed statistics add %d\\n", stack_id);
+    return;
+  }
+  cinfo = *existing_cinfo;
+
+  if (cinfo.number_of_allocs > 0)
+    cinfo.number_of_allocs--;
+
+  if (survivor_time_ns > cinfo.last_longest_survivor) {
+    cinfo.last_longest_survivor = survivor_time_ns;
+  }
+
+  combined_allocs_hash.update(&stack_id, &cinfo);
+}
+
+static inline int fd_close_enter(unsigned fd) {
+  u64 pidfd = bpf_get_current_pid_tgid() & 0xffffffff00000000L;
+  struct alloc_info_t *info;
+  u64 survivor_time_ns = bpf_ktime_get_ns();
+
+  pidfd += (u32)fd;
+
+  info = fd_allocs_hash.lookup(&pidfd);
+  if (info == 0) {
+    return 0;
+  }
+
+  survivor_time_ns -= info->timestamp_ns;
+  ##PRINT_DEBUG##
+  bpf_trace_printk("[%s-%u] free fd[%d] entered \\n",
+                    info->comm, info->tid, fd);
+
+  fd_allocs_hash.delete(&pidfd);
+  update_statistics_del(info->stack_id, survivor_time_ns);
+
+  return 0;
+}
+
+// close fd
+// https://github.com/torvalds/linux/commit/8760c909f54a82aaa6e76da19afe798a0c77c3c3
+// file: Rename __close_fd to close_fd and remove the files parameter
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)
+int kprobe____close_fd(struct pt_regs *ctx, unsigned fd)
+#else
+int kprobe__close_fd(struct pt_regs *ctx, unsigned int fd)
+#endif
+{
+  ##FILTER_PID##
+  FD_MONITOR_TRIGGER(fd)
+  return fd_close_enter(fd);
+}
+
+// alloc_fd
+int kretprobe__get_unused_fd_flags(struct pt_regs *ctx)
+{
+  ##FILTER_PID##
+  int fd = PT_REGS_RC(ctx);
+  FD_MONITOR_TRIGGER(fd)
+
+  return fd_alloc_return(ctx);
+}
+"""
+# parse arguments
+if args.pid != -1:
+    bpf_text = bpf_text.replace('##FILTER_PID##',
+        'if (bpf_get_current_pid_tgid() >> 32 != %s) { return 0; }' % args.pid)
+    print("start trace pid= %s" % args.pid)
+else:
+    bpf_text = bpf_text.replace('##FILTER_PID##', '')
+
+interval = args.interval
+min_allow_survival = args.min_allow
+
+if args.debug:
+    bpf_text = bpf_text.replace("##PRINT_DEBUG##", "if(1)")
+else:
+    bpf_text = bpf_text.replace("##PRINT_DEBUG##", "if(0)")
+
+if args.ebpf:
+    print(bpf_text)
+
+# load c text
+bpf = BPF(text=bpf_text)
+
+def get_pname(id):
+    comm_path=("/proc/%d/comm"%id)
```

Comment:
Did you test your code with pep8 for the coding style?
https://github.com/iovisor/bcc/blob/master/CONTRIBUTING-SCRIPTS.md => Step 12.

aaa.py:193:19: E226 missing whitespace around arithmetic operator
        t = t_name+"-"+str(tid)